File: 733-flood-fill/733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, newColor) {
  const oldColor = image[sr][sc]
 fillColor(image, sr, sc, oldColor, newColor)
    return image
}

const fillColor = (image, x ,y, oldColor, newColor) => {
    const rowInbound = 0 <= x && x <image.length
    const colInbound = 0 <= y && y <image[0].length
    
    if (!rowInbound || !colInbound || image[x][y] !== oldColor || image[x][y] === newColor) return 
    image[x][y] = newColor
    
    fillColor(image, x - 1, y, oldColor, newColor)
    fillColor(image, x + 1, y, oldColor, newColor)
    fillColor(image, x, y + 1, oldColor, newColor)
    fillColor(image, x, y - 1, oldColor, newColor)
    
}

File: 733-flood-fill/733-flood-fill.py
def ffill(old_color, new_color, i, j, image):

    if i < len(image) and i >= 0 and j < len(image[0]) and j >= 0 and image[i][j] == old_color:

        image[i][j] = new_color
        ffill(old_color, new_color, i, j+1, image)
        ffill(old_color, new_color, i+1, j, image)
        ffill(old_color, new_color, i-1, j, image)
        ffill(old_color, new_color, i, j-1, image)

    return image

class Solution:
    def floodFill(self, image, sr, sc, color):
        if image[sr][sc] == color:
            return image
        
        return ffill(image[sr][sc], color, sr, sc, image)


            